### PR TITLE
Stm32 g4 variant

### PR DESCRIPTION
**Pull request by ABOSTM**
**Summary**

Stm32 g4 variant

### CODE REVIEW

**Comment by ABOSTM** [comment]
@fpistm Review requested for last commit:
    [variant] Add Generic G4 and Nucleo G4

File: README.md [hunk 1]
```diff
@@ -240,6 +241,14 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
 | :yellow_heart: | STM32G031K4<br>STM32G031K6<br>STM32G031K8 | Generic Board | **2.0.0** |  |
 | :yellow_heart: | STM32G071R6<br>STM32G071R8<br>STM32G071RB | Generic Board | **2.0.0** |  |
 
+### Generic STM32G4 boards
+
+| Status | Device(s) | Name | Release | Notes |
+| :----: | :-------: | ---- | :-----: | :---- |
+| :yellow_heart: | STM32G474RE | Generic Board | **2.0.0** |  |
```
Comment (fpistm) [inline README.md]: The generic provide also support of STM32G474RB and STM32G474RC

File: README.md [hunk 2]
```diff
@@ -240,6 +241,14 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
...
+| :yellow_heart: | STM32G431RB | Generic Board | **2.0.0** |  |
```
Comment (fpistm) [inline README.md]: Ditto 6 and 8

File: README.md [hunk 3]
```diff
@@ -240,6 +241,14 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
...
+| :yellow_heart: | STM32G431KB | Generic Board | **2.0.0** |  |
```
Comment (fpistm) [inline README.md]: ditto 6 and 8

File: boards.txt [hunk 4]
```diff
@@ -351,7 +351,7 @@ Nucleo_64.menu.pnum.NUCLEO_G431RB.build.flags.fp=-mfpu=fpv4-sp-d16 -mfloat-abi=h
 Nucleo_64.menu.pnum.NUCLEO_G431RB.build.board=NUCLEO_G431RB
 Nucleo_64.menu.pnum.NUCLEO_G431RB.build.series=STM32G4xx
 Nucleo_64.menu.pnum.NUCLEO_G431RB.build.product_line=STM32G431xx
-Nucleo_64.menu.pnum.NUCLEO_G431RB.build.variant=STM32G4xx/NUCLEO_G431RB
+Nucleo_64.menu.pnum.NUCLEO_G431RB.build.variant=STM32G4xx/Generic_G431RB
```
Comment (fpistm) [inline boards.txt]: Generic_G431Rx

File: boards.txt [hunk 5]
```diff
@@ -364,7 +364,7 @@ Nucleo_64.menu.pnum.NUCLEO_G474RE.build.flags.fp=-mfpu=fpv4-sp-d16 -mfloat-abi=h
 Nucleo_64.menu.pnum.NUCLEO_G474RE.build.board=NUCLEO_G474RE
 Nucleo_64.menu.pnum.NUCLEO_G474RE.build.series=STM32G4xx
 Nucleo_64.menu.pnum.NUCLEO_G474RE.build.product_line=STM32G474xx
-Nucleo_64.menu.pnum.NUCLEO_G474RE.build.variant=STM32G4xx/NUCLEO_G474RE
+Nucleo_64.menu.pnum.NUCLEO_G474RE.build.variant=STM32G4xx/Generic_G474RE
```
Comment (fpistm) [inline boards.txt]: Generic_G474Rx

File: boards.txt [hunk 6]
```diff
@@ -537,6 +537,19 @@ Nucleo_32.menu.pnum.NUCLEO_G031K8.build.variant=STM32G0xx/Generic_G031Kx
...
+Nucleo_32.menu.pnum.NUCLEO_G431KB=Nucleo G431KB
+Nucleo_32.menu.pnum.NUCLEO_G431KB.node="NODE_G431KB,NOD_G431KB"
+Nucleo_32.menu.pnum.NUCLEO_G431KB.upload.maximum_size=131072
+Nucleo_32.menu.pnum.NUCLEO_G431KB.upload.maximum_data_size=32768
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.mcu=cortex-m4
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.flags.fp=-mfpu=fpv4-sp-d16 -mfloat-abi=hard
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.board=NUCLEO_G431KB
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.series=STM32G4xx
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.product_line=STM32G431xx
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.variant=STM32G4xx/Generic_G431KB
```
Comment (fpistm) [inline boards.txt]: Generic_G431Kx

File: boards.txt [hunk 7]
```diff
@@ -1976,6 +1976,60 @@ GenG0.menu.upload_method.dfuMethod.upload.protocol=2
...
+GenG4.menu.pnum.GENERIC_G431RB.build.product_line=STM32G431xx
+GenG4.menu.pnum.GENERIC_G431RB.build.variant=STM32G4xx/Generic_G431RB
+
+# Generic G474RE board
+GenG4.menu.pnum.GENERIC_G474RE=Generic G474RE
+GenG4.menu.pnum.GENERIC_G474RE.upload.maximum_size=524288
+GenG4.menu.pnum.GENERIC_G474RE.upload.maximum_data_size=131072
+GenG4.menu.pnum.GENERIC_G474RE.build.board=GENERIC_G474RE
+GenG4.menu.pnum.GENERIC_G474RE.build.product_line=STM32G474xx
+GenG4.menu.pnum.GENERIC_G474RE.build.variant=STM32G4xx/Generic_G474RE
```
Comment (fpistm) [inline boards.txt]: It misses some generic variant for each G4

File: boards.txt [hunk 8]
```diff
@@ -2038,7 +2092,25 @@ GenH7.menu.upload_method.dfuMethod.upload.options=-g
...
+GenG0.build.extra_flags=-D{build.product_line} {build.enable_usb} {build.xSerial} -D__CORTEX_SC=0
+
+# Generic G030K6
+GenG0.menu.pnum.GENERIC_G030K6=Generic G030K6
+GenG0.menu.pnum.GENERIC_G030K6.upload.maximum_size=32768
+GenG0.menu.pnum.GENERIC_G030K6.upload.maximum_data_size=8192
+GenG0.menu.pnum.GENERIC_G030K6.build.board=GENERIC_G030K6
+GenG0.menu.pnum.GENERIC_G030K6.build.product_line=STM32G030xx
+GenG0.menu.pnum.GENERIC_G030K6.build.variant=STM32G0xx/Generic_G030Kx
+
```
Comment (fpistm) [inline boards.txt]: ??

File: variants/STM32G4xx/Generic_G431KB/variant.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ */
+#if defined(ARDUINO_GENERIC_G431KB)
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431KB/variant.cpp]: #if defined(ARDUINO_GENERIC_G431K6)  || defined(ARDUINO_GENERIC_G431K8)  || defined(ARDUINO_GENERIC_G431KB)

File: variants/STM32G4xx/Generic_G431KB/variant.cpp [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  23, // A9,  PF0
+  24  // A10, PF1
+};
+
+/******************************************************************************/
+/*            PLL (clocked by HSE) used as System clock source                */
+/******************************************************************************/
+static bool SetSysClock_PLL_HSE(bool bypass)
+
+
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431KB/variant.cpp]: to remove

File: variants/STM32G4xx/Generic_G431KB/variant.h [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ */
+#if !defined(ARDUINO_GENERIC_G431KB)
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431KB/variant.h]: Ditto missing some defined

File: variants/STM32G4xx/Generic_G431KB/variant_NUCLEO_G431KB.cpp [hunk 12]
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431KB/variant_NUCLEO_G431KB.cpp]: 2019-2021

File: variants/STM32G4xx/Generic_G431KB/variant_NUCLEO_G431KB.h [hunk 13]
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431KB/variant_NUCLEO_G431KB.h]: 2019-2021

File: variants/STM32G4xx/Generic_G431RB/variant.cpp [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ */
+#if defined(ARDUINO_GENERIC_G431RB)
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431RB/variant.cpp]: ditto

File: variants/STM32G4xx/Generic_G431RB/variant.cpp [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+      }
+    }
+  }
+  return ret;
+}
+
+WEAK void SystemClock_Config(void)
+{
+  /*
+   * HSE_VALUE defined in the stm32g0xx_hal_conf_default.h is used.
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431RB/variant.cpp]: stm32g4xx_hal_conf_default.h

File: variants/STM32G4xx/Generic_G431RB/variant.h [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ */
+#if !defined(ARDUINO_GENERIC_G431RB)
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431RB/variant.h]: ditto

File: variants/STM32G4xx/Generic_G431RB/variant_NUCLEO_G431RB.cpp [hunk 17]
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431RB/variant_NUCLEO_G431RB.cpp]: 2019-2021

File: variants/STM32G4xx/Generic_G431RB/variant_NUCLEO_G431RB.h [hunk 18]
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G431RB/variant_NUCLEO_G431RB.h]: 2019-2021

File: variants/STM32G4xx/Generic_G474RE/variant.cpp [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ */
+#if defined(ARDUINO_GENERIC_G474RE)
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G474RE/variant.cpp]: ditto

File: variants/STM32G4xx/Generic_G474RE/variant.h [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ */
+#if !defined(ARDUINO_GENERIC_G474RE)
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G474RE/variant.h]: ditto

File: variants/STM32G4xx/Generic_G474RE/variant_NUCLEO_G474RE.cpp [hunk 21]
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G474RE/variant_NUCLEO_G474RE.cpp]: 2019-2021

File: variants/STM32G4xx/Generic_G474RE/variant_NUCLEO_G474RE.h [hunk 22]
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
Comment (fpistm) [inline variants/STM32G4xx/Generic_G474RE/variant_NUCLEO_G474RE.h]: 2019-2021

**Review by fpistm — CHANGES_REQUESTED** [review]

**Comment by ABOSTM** [comment]
@fpistm ,
Rebase on latest automatic generic variant generation.
Fixes CI issue (Astyle + compilation issue)
Take your remarks into account

**Comment by ABOSTM** [comment]
ABOSTM fork. Close this PR

**Pull request closed without merging**